### ISSUE TITLE
Gradient-based optimization for geometric changes

### ISSUE THREAD

**Issue by ymatthew2466** [mitsuba-renderer/mitsuba3#783]
## Summary

Good afternoon. I'm new to graphics, so please excuse any ambiguity in my description. I'm having trouble using gradient-based optimization to recover the original image after a modification. 

## System configuration

OS: 12.5.1
  CPU: b'Apple M1 Pro'
  Python: 3.11.4 (main, Jun  7 2023, 00:34:23) [Clang 14.0.0 (clang-1400.0.29.202)]
  LLVM: 16.0.5

  Dr.Jit: 0.4.2
  Mitsuba: 3.3.0
     Is custom build? False
     Compiled with: AppleClang 14.0.0.14000029
     Variants:
        scalar_rgb
        scalar_spectral
        llvm_ad_rgb

## Description

I followed the tutorial for object pose estimation to create an identity transformation on a sphere. 
```
opt = mi.ad.Adam(lr=0.05)
opt['sphere1.translation'] = mi.Vector3f(0,0,0)
opt['sphere1.axis'] = mi.Vector3f(0.0,0.0,1.0)
opt['sphere1.angle'] = mi.Float(0.0)
opt['sphere1.scale'] = mi.Vector3f(1,1,1)
```
Then, I translated and scaled the sphere in Photoshop and passed that image as a TensorXf object. When running the optimization loop to recover the original sphere from the Photoshopped image, I noticed that the MSE was constant throughout, even though the output image was altered. I'm curious whether there's an error in my implementation.

Here is my code:
```
import drjit as dr
import mitsuba as mi
import matplotlib.pyplot as plt
def imshow( image ):
	plt.imshow(image)
	plt.axis('off');
	plt.show()

mi.set_variant('llvm_ad_rgb')

scene = mi.load_file('scenes/cbox.xml', res=256, integrator='prb')
render_bitmap = mi.Bitmap('render1transform.png').convert(mi.Bitmap.PixelFormat.RGB, mi.Struct.Type.Float32, srgb_gamma=True)

image_ref = mi.TensorXf(render_bitmap)
mi.util.write_bitmap("render1.png", image_ref)
params = mi.traverse(scene)

keys = ['3sphere.to_world']
print(type(params[keys[0]]))

param_refs = [] 
for key in keys:
	param_refs.append(params[key])

image_init = mi.render(scene, spp=128)
mi.util.write_bitmap("render1.png", image_init)
imshow( mi.util.convert_to_bitmap(image_init) )

opt = mi.ad.Adam(lr=0.05)
opt['sphere1.translation'] = mi.Vector3f(0,0,0)
opt['sphere1.axis'] = mi.Vector3f(0.0,0.0,1.0)
opt['sphere1.angle'] = mi.Float(0.0)
opt['sphere1.scale'] = mi.Vector3f(1,1,1)

def apply_transformation(params, opt):
    T = ( mi.Transform4f
	     .translate(opt['sphere1.translation'])
		 .rotate(opt['sphere1.axis'], opt['sphere1.angle'])
		 .scale(opt['sphere1.scale'])
		 )

    params['3sphere.to_world'] = T
    params.update()

def mse(image):
	return dr.mean(dr.sqr(image - image_ref))

iteration_count = 50

errors = []
for it in range(iteration_count):
	apply_transformation( params, opt )
	image = mi.render(scene, params, spp=4) # render a rough sketch
	loss = mse(image) # calc loss by mean square error
	dr.backward(loss) # magic inverse rendering
	opt.step() # gradient step

	print(f"Iteration {it:02d}: ", loss, end='\r')
	errors.append(loss)
print('\nOptimization complete.')

image_final = mi.render(scene, spp=512)
imshow( mi.util.convert_to_bitmap(image_final) )

plt.plot(errors)
plt.xlabel('Iteration'); plt.ylabel('MSE(param)'); plt.title('Parameter error plot');
plt.show()
```

**Comment by njroussel** [comment]
Hi @ymatthew2466 

At the very top of the tutorial you mentioned, there is a section about the choice of integrator for such a task. Moving geometry will create visibility discontinuities which can only be handled by certain integrators in Mitsuba currently. These are integrators that end with the `*_reparam` suffix.

**Comment by ymatthew2466** [comment]
Thanks for the response @njroussel. Is it possible to use these ``*_reparam`` integrators by editing the XML file? Documentation suggests that it's only doable when loading a scene with mi.load_dict

https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_integrators.html#reparameterized-direct-integrator-direct-reparam

**Comment by njroussel** [comment]
I guess you found your answer, but just in case:

Plugins which are defined in Python such as `prb` or `prb_reparam` cannot be used when rendering through the Mitsuba CLI. However, it is still possible to load these plugins from a XML description, but it must be done in Python using `mi.load_file` or `mi.load_string`.
